### PR TITLE
auto mTLS: Change label name to tlsMode

### PR DESCRIPTION
**Pull request by rshriram**
Most of this PR is simply renaming mtlsReady to tlsMode, so that people can add their own modes and write EnvoyFilters to update the transport socket matchers.

Signed-off-by: Shriram Rajagopalan <rshriram@tetrate.io>

### CODE REVIEW

**Comment by rshriram** [comment]
cc @incfly ..

File: pilot/pkg/model/service.go
Comment (incfly) [inline pilot/pkg/model/service.go]: I think `istio_mutual` is better than `istio`. mtls can not be assumed, if we're introducing more and more tls config mode.

File: pilot/pkg/model/service.go
Comment (incfly) [inline pilot/pkg/model/service.go]: comment is no longer accurate. `mTLS autopilot` now you use this label for other cases more than mTLS..

File: pilot/pkg/model/service.go
Comment (rshriram) [reply pilot/pkg/model/service.go]: We don’t want a long label for each endpoint do we ?

File: pkg/kube/inject/webhook.go
Comment (hzxuzhonghu) [inline pkg/kube/inject/webhook.go]: invalid value

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (diemtvu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: We do need trusted_ca to validate server cert, right?

**Review by diemtvu — COMMENTED** [review]
To be clear, it's not change the fallback, but add more options for socket match beside mTLS on and off, right?

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (rshriram) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: No. that beats the purpose of automatic fallback. people can then define dest rules if they want

File: pilot/pkg/model/service.go
Comment (rshriram) [reply pilot/pkg/model/service.go]: @incfly wanted simple-no-verify because we are not verifying server cert

File: pilot/pkg/serviceregistry/external/conversion_test.go
Comment (rshriram) [reply pilot/pkg/serviceregistry/external/conversion_test.go]: need to fix this test

**Comment by rshriram** [comment]
> To be clear, it's not change the fallback, but add more options for socket match beside mTLS on and off, right?

@diemtvu this is to change the fallback. Not everyone likes a fallback to clear text, even in permissive mode, especially when one-way https is already supported in most organizations

**Review by esnible — APPROVED** [review]
The _istioctl_ changes are approved.  I did not look hard at the other changes.

File: pilot/pkg/model/service.go
Comment (rshriram) [reply pilot/pkg/model/service.go]: The impact is same as not specified. But it still allows the user to explicitly add that label. In other words, this is same as `security.istio.io/mtlsReady: false` [which is same as not specified]. In this case, its `security.istio.io/tlsMode: istio|simple-no-verify|disabled` [pick one]

File: pilot/pkg/model/service.go
Comment (incfly) [reply pilot/pkg/model/service.go]: sure... good point.

File: pilot/pkg/model/service.go
Comment (incfly) [reply pilot/pkg/model/service.go]: right, there's a behavior difference compared with TLSSettings_simple

**Comment by diemtvu** [comment]
> > To be clear, it's not change the fallback, but add more options for socket match beside mTLS on and off, right?
> 
> @diemtvu this is to change the fallback. Not everyone likes a fallback to clear text, even in permissive mode, especially when one-way https is already supported in most organizations

Technically speaking, in this PR, you are adding a third match rule for the new label. The fallback is still plaintext when no label available.

My concern is the label now become some sort of API. Is this better to actual define a "fallback" TLS options instead? (e.g in DR)

**Comment by rshriram** [comment]
>My concern is the label now become some sort of API. Is this better to actual define a "fallback" TLS options instead? (e.g in DR)

Dont look at this from implementation point of view. Look at it from a user point of view: If there is mTLS we will use it, if there is no mtls, we look for simple-tls - if present we will pick simple TLs, and if not present, then we will go to plaintext.

**Comment by diemtvu** [comment]
> > My concern is the label now become some sort of API. Is this better to actual define a "fallback" TLS options instead? (e.g in DR)
> 
> Dont look at this from implementation point of view. Look at it from a user point of view: If there is mTLS we will use it, if there is no mtls, we look for simple-tls - if present we will pick simple TLs, and if not present, then we will go to plaintext.

The different is users now interact with the label to get to the configuration they want. The original mTLS label is more transparent to users. 

If we go down this way, then I think we should allow users to define the (socket) configuration that match with their choice of label, which allow them to define their own settings, including trusted ca or secure naming. We can change the label name and value in this PR first, but may hold up the addition for new socket match until later?

File: pilot/pkg/model/service.go
Comment (hzxuzhonghu) [reply pilot/pkg/model/service.go]: ic

**Comment by rshriram** [comment]
>The different is users now interact with the label to get to the configuration they want. The original mTLS label is more transparent to users.

Remember that in VM world, there is no transparency in either case (old or new proposed label). In kubernetes world, there is no change (most users only use mtls and the label is automatically injected, and if not present, we fallback to plain text as usual). In VMs, we have to manually author service entries with this label. So while adding, it is easier to specify what mode is desired.

**Comment by rshriram** [comment]
@incfly / @diemtvu / @linsun and @istio/wg-networking-maintainers-pilot I have changed the scope of this PR. This currently just changes the label name from mtlsReady to tlsMode. However, the values recognized by the code is still same old `istio` and `disabled` or no label at all. Nothing changes on that front. This label will be pushed to all endpoints regardless of the value (istio|disabled|any-custom-value). This allows end users to write EnvoyFilters to add custom transport socket matchers.

File: pilot/pkg/model/service.go
Comment (diemtvu) [inline pilot/pkg/model/service.go]: nit. May be returning an enum instead.

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (diemtvu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: I'm not quite understand what the differences between these 2. Can the plaintext-nolabel can be used for both no-label and disable-label case?

**Comment by howardjohn** [comment]
Looks fine to me, but will let others that have been working on this take a look

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (GregHanson) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: Out of curiosity, are there any upcoming uses/features/design plans for TransportSocketMatches in upstream envoy where `istio` would not be an appropriate name?  i.e. if we as istio developers decide to append additional TransportSocketMatches to cover additional features, `istio` may not be appropriate since we would be appending more than one Istio-themed socket match.  My assumption is these cluster settings will remain pretty invisible to users so changing names down the line shouldn't be breaking from a customer standpoint correct?

File: pilot/pkg/networking/util/util.go
Comment (GregHanson) [inline pilot/pkg/networking/util/util.go]: @howardjohn raised a performance concern with initializing this struct/map constantly [see here](https://github.com/istio/istio/pull/17621#discussion_r335547832), but I guess this can't be avoided since the key's value is dynamic rather than always `true`?

**Review by GregHanson — APPROVED** [review]

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (linsun) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: my guess is plaintext-disabled-label is for one way TLS but seems the code isn't updated properly?

**Review by linsun — APPROVED** [review]
I saw @diemtvu has one or two concerns but nothing major.

File: pilot/pkg/model/service.go
Comment (rshriram) [reply pilot/pkg/model/service.go]: See the explanation below for why we are returning the string. This is to let others extend the modes and use EnvoyFilter for customization. If you change this to an enum, then it wont work.

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (rshriram) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: No.. plaintext disabled is same as label not specified. But i thought I would be explicit

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (rshriram) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: So this is just name of object and not the type. We have to provide some name and this is mostly used for stats or just logging.

File: pilot/pkg/networking/util/util.go
Comment (rshriram) [reply pilot/pkg/networking/util/util.go]: Yes. And the goal of this is to allow end users to provide any value for the key and not just the specific values. By allowing them to provide any value, they can then use this key-value to create otehr custom transport socket matchers (using EnvoyFilter). Essentially, in this case, we are trading off performance for flexibility.

**Comment by incfly** [comment]
> This label will be pushed to all endpoints regardless of the value (istio|disabled|any-custom-value). This allows end users to write EnvoyFilters to add custom transport socket matchers.

Does Envoy has the API to override upstream `transport_socket_matches`? I know we can do it for ALPN in the `tls_context`, but not aware that we can do full fledged things for all upstream cluster configuration.

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (diemtvu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: Not very convinced that we need both, but it's a minor detail though.

**Review by diemtvu — APPROVED** [review]

File: install/kubernetes/helm/istio/charts/mixer/templates/deployment.yaml
Comment (rshriram) [reply install/kubernetes/helm/istio/charts/mixer/templates/deployment.yaml]: are you sure I can remove this? I am simply renaming labels and not removing them, so that I dont get caught up in another issue. If this is sure to work, then I can remove the label

Comment (incfly) [reply install/kubernetes/helm/istio/charts/mixer/templates/deployment.yaml]: up to you. i have another PR in flight to remove this anyway.

File: pilot/pkg/networking/util/util.go
Comment (howardjohn) [reply pilot/pkg/networking/util/util.go]: I think its fine, if we really need to we can add a cache per tlsMode, but probably premature optimization

**Review by howardjohn — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
/retest

I have seen TestProcessor_Publishing failure many times.

File: install/kubernetes/helm/istio/charts/mixer/templates/deployment.yaml
Comment (GregHanson) [reply install/kubernetes/helm/istio/charts/mixer/templates/deployment.yaml]: @incfly the issue arises when `--set global.controlPlaneSecurityEnabled=true` and automtls is enabled.  Please make sure this is tested in your other PR

**Pull request merged**
